bbb/meta-bbb/recipes-core/trebletrouble/files/src/audio_recorder.c

```diff
@@ -0,0 +1,212 @@
+  }
+
+  /* Two channels (stereo) */
+  err = snd_pcm_hw_params_set_channels(handle, params, hdr -> number_of_channels);
+  if (err) {
+    fprintf(stderr, "Error setting channels: %s\n", snd_strerror(err));
+    snd_pcm_close(handle);
+    return err;
+  }
+
+  /* 44100 bits/second sampling rate (CD quality) */
+  sampleRate = hdr->sample_rate;
+  err = snd_pcm_hw_params_set_rate_near(handle, params, &sampleRate, &dir);
+  if (err) {
+    fprintf(stderr, "Error setting sampling rate (%d): %s\n", sampleRate, snd_strerror(err));
+    snd_pcm_close(handle);
+    return err;
+  }
+  hdr -> sample_rate = sampleRate;
+
+  /* Set period size */
+  err = snd_pcm_hw_params_set_period_size_near(handle, params, &frames, &dir);
+  if (err) {
+    fprintf(stderr, "Error setting period size: %s\n", snd_strerror(err));
+    snd_pcm_close(handle);
+    return err;
+  }
+
+  /* Write the parameters to the driver */
+  err = snd_pcm_hw_params(handle, params);
+  if (err < 0) {
+    fprintf(stderr, "Unable to set HW parameters: %s\n", snd_strerror(err));
+    snd_pcm_close(handle);
+    return err;
+  }
+
+  /* Use a buffer large enough to hold one period */
+  err = snd_pcm_hw_params_get_period_size(params, &frames, &dir);
+  if (err){
+    fprintf(stderr, "Error retrieving period size: %s\n", snd_strerror(err));
+    snd_pcm_close(handle);
+    return err;
+  }
+
+  size = frames * hdr-> bits_per_sample / 8 * hdr -> number_of_channels; // 2 bytes/sample, 2 channels
+  buffer = (char *) malloc(size);
+  if (!buffer) {
+    fprintf(stdout, "Buffer error.\n");
+    snd_pcm_close(handle);
+    return -1;
+  }
+
+  err = snd_pcm_hw_params_get_period_time(params, &sampleRate, &dir);
+  if (err) {
+    fprintf(stderr, "Error retrieving period time: %s\n", snd_strerror(err));
+    snd_pcm_close(handle);
+    free(buffer);
+    return err;
+  }
+
+  uint32_t pcm_data_size = hdr-> sample_rate * hdr-> bytes_per_frame * (duration/1000);
+  hdr -> file_size = pcm_data_size + 36;
+
+  filedesc = open(fileName, O_WRONLY | O_CREAT, 0644);
+  err = writeWAVHeader(filedesc, hdr);
+  if (err) {
+    fprintf(stderr, "Error writing .wav header.");
+    snd_pcm_close(handle);
+    free(buffer);
+    close(filedesc);
+    return err;
+  }
+
+  int totalFrames = 0;
+
+  int i = ((duration * 1000) / (hdr->sample_rate / frames));
+  for(i; i > 0; i--) {
+    err = snd_pcm_readi(handle, buffer, frames);
+    totalFrames += err;
+    if (err == -EPIPE) {
+      fprintf(stderr, "Overrun occurred: %d\n", err);
+    }
+    if (err < 0) {
+      snd_pcm_recover(handle, err, 0);
+    }
+    // Still an error, need to exit.
+    if (err < 0) {
+      fprintf(stderr, "Error occurred while recording: %s\n", snd_strerror(err));
+      snd_pcm_close(handle);
+      free(buffer);
+      close(filedesc);
+      return err;
+    }
+    write(filedesc, buffer, size);
+  } // end for loop
+  
+  close(filedesc);
+  snd_pcm_drain(handle);
+  snd_pcm_close(handle);
+  free(buffer);
+  return 0;
+}
```

Comment:
Add the goto here. You should free resources in the opposite order in which you allocated them (think of them as a stack) so that your goto labels will fall-through one another.